machineconfiguration/v1/types.go

```diff
@@ -0,0 +1,389 @@
+
+// MCOConfig describes configuration for MachineConfigOperator.
+type MCOConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +required
+	Spec MCOConfigSpec `json:"spec"`
+}
+
+// MCOConfigSpec is the spec for MCOConfig resource.
+type MCOConfigSpec struct {
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MCOConfigList is a list of MCOConfig resources
+type MCOConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MCOConfig `json:"items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ControllerConfig describes configuration for MachineConfigController.
+// This is currently only used to drive the machineconfigs generated by the TemplateController.
+type ControllerConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ControllerConfigSpec   `json:"spec"`
+	Status ControllerConfigStatus `json:"status"`
+}
+
+// ControllerConfigSpec is the spec for ControllerConfig resource.
+type ControllerConfigSpec struct {
+	ClusterDNSIP        string `json:"clusterDNSIP"`
+	CloudProviderConfig string `json:"cloudProviderConfig"`
+
+	// The openshift platform, e.g. "libvirt", "openstack", "aws", or "none"
+	Platform string `json:"platform"`
+
+	EtcdDiscoveryDomain string `json:"etcdDiscoveryDomain"`
+
+	// CAs
+	EtcdCAData       []byte `json:"etcdCAData"`
+	EtcdMetricCAData []byte `json:"etcdMetricCAData"`
+	RootCAData       []byte `json:"rootCAData"`
+
+	// PullSecret is the default pull secret that needs to be installed
+	// on all machines.
+	PullSecret *corev1.ObjectReference `json:"pullSecret,omitempty"`
+
+	// Images is map of images that are used by the controller.
+	Images map[string]string `json:"images"`
+
+	// Sourced from configmap/machine-config-osimageurl
+	OSImageURL string `json:"osImageURL"`
+}
+
+// ControllerConfigStatus is the status for ControllerConfig
+type ControllerConfigStatus struct {
+	// The generation observed by the controller.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Represents the latest available observations of current state.
+	// +optional
+	Conditions []ControllerConfigStatusCondition `json:"conditions,omitempty"`
+}
+
+// ControllerConfigStatusCondition contains condition information for ControllerConfigStatus
+type ControllerConfigStatusCondition struct {
+	// type specifies the state of the operator's reconciliation functionality.
+	Type ControllerConfigStatusConditionType `json:"type"`
+
+	// status of the condition, one of True, False, Unknown.
+	Status corev1.ConditionStatus `json:"status"`
+
+	// lastTransitionTime is the time of the last update to the current status object.
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
```

Comment:
omitempty has no effect for `metav1.Time`.

Reply:
consider `// +nullable` as a zero value time will render as such.

Reply:
We have LastTransitionTime in 4 places. Adding `// +nullable` on all of them.